app/code/Magento/QuoteBundleOptions/Model/Cart/BuyRequest/BundleDataProvider.php

```diff
@@ -0,0 +1,75 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\QuoteBundleOptions\Model\Cart\BuyRequest;
+
+use Magento\Framework\Exception\LocalizedException;
+use Magento\Quote\Model\Cart\BuyRequest\BuyRequestDataProviderInterface;
+use Magento\Quote\Model\Cart\Data\CartItem;
+
+/**
+ * Data provider for bundle product buy requests
+ */
+class BundleDataProvider implements BuyRequestDataProviderInterface
+{
+    private const OPTION_TYPE = 'bundle';
+
+    /**
+     * @inheritdoc
+     *
+     * @throws LocalizedException
+     */
+    public function execute(CartItem $cartItem): array
+    {
+        $bundleOptionsData = [];
+
+        foreach ($cartItem->getSelectedOptions() as $optionData) {
+            // phpcs:ignore Magento2.Functions.DiscouragedFunction
+            $optionData = \explode('/', base64_decode($optionData->getId()));
```

Comment:
Hmm. I don't think we should be doing this. An ID is something that should be opaque and there could be several ways we arrive at an ID (base64 is just one of it). And we should not be decoding it to get optionId,  optionValueId etc. It should be explicit from the clients. And this decode is used on almost every data provider on this PR. @paliarush @DrewML is this ok?

Reply:
As soon as no decoding/encoding is required on the storefront - the approach follows the design.

Reply:
Let me try to explain this better. Consider the case for a bundle product with 2 options.
Option 1 has 2 products, with **dropdown** type and **user can enter quantity**
Option 2 has 2 products with **multiselect** type with a fixed quantity

So the way we are encoding the id for each product in the option is base_64(bundle/<option-id>/<option-value-id>/<option-quantity>). It takes in the hypothesis that option_quantity is always fixed, which is not since user can change it.

**Problem 1**
When we add the bundle item to the cart, the design is not respecting the quantity the user is passing, rather its decoding from the encoded id. The user entered quantities will be ignored here.

*Problem 2*
<img width="944" alt="Screen Shot 2020-07-14 at 5 22 48 PM" src="https://user-images.githubusercontent.com/43612864/87482306-c969c400-c5f6-11ea-8461-7b4b6afa5a6d.png">

For a configurable product the clients have the code and value_index. And I am not sure why we are encoding them and also returning the non encoded values at the same time. Might be to support the old mutations? 

It is evident that we are encoding the ids for entities/types with very specific properties of each product type. More than cache just keys it will have business/behavioral implications. If we are going to rely on the ids for decoding and using the properties I think we should have strict interfaces per product type for encoding/decoding and also interface for abstracting the algorithm we use for the same. Currently its all over the place.